### ISSUE TITLE
Build failure due to CMake "file COPY cannot copy file"

### ISSUE THREAD

**Issue by MilhouseVH** [xbmc/xbmc#16000]
Kodi frequently fails with:
```
[7/1529] No update step for 'dvdcss'
[8/1529] Generating build/xbmc/CompileInfo.cpp, addons/audioencoder.kodi.builtin.aac/addon.xml, addons/audioencoder.kodi.builtin.wma/addon.xml, addons/kodi.binary.global.audioengine/addon.xml, addons/kodi.binary.global.filesystem/addon.xml, addons/kodi.binary.global.general/addon.xml, addons/kodi.binary.global.gui/addon.xml, addons/kodi.binary.global.main/addon.xml, addons/kodi.binary.global.network/addon.xml, addons/kodi.binary.instance.audiodecoder/addon.xml, addons/kodi.binary.instance.audioencoder/addon.xml, addons/kodi.binary.instance.game/addon.xml, addons/kodi.binary.instance.imagedecoder/addon.xml, addons/kodi.binary.instance.inputstream/addon.xml, addons/kodi.binary.instance.peripheral/addon.xml, addons/kodi.binary.instance.pvr/addon.xml, addons/kodi.binary.instance.screensaver/addon.xml, addons/kodi.binary.instance.vfs/addon.xml, addons/kodi.binary.instance.videocodec/addon.xml, addons/kodi.binary.instance.visualization/addon.xml, addons/xbmc.addon/addon.xml
FAILED: build/xbmc/CompileInfo.cpp addons/audioencoder.kodi.builtin.aac/addon.xml addons/audioencoder.kodi.builtin.wma/addon.xml addons/kodi.binary.global.audioengine/addon.xml addons/kodi.binary.global.filesystem/addon.xml addons/kodi.binary.global.general/addon.xml addons/kodi.binary.global.gui/addon.xml addons/kodi.binary.global.main/addon.xml addons/kodi.binary.global.network/addon.xml addons/kodi.binary.instance.audiodecoder/addon.xml addons/kodi.binary.instance.audioencoder/addon.xml addons/kodi.binary.instance.game/addon.xml addons/kodi.binary.instance.imagedecoder/addon.xml addons/kodi.binary.instance.inputstream/addon.xml addons/kodi.binary.instance.peripheral/addon.xml addons/kodi.binary.instance.pvr/addon.xml addons/kodi.binary.instance.screensaver/addon.xml addons/kodi.binary.instance.vfs/addon.xml addons/kodi.binary.instance.videocodec/addon.xml addons/kodi.binary.instance.visualization/addon.xml addons/xbmc.addon/addon.xml 
cd /var/lib/jenkins/LE/build2/workspace/box-trn9/build.LibreELEC-RK3328.arm-9.1-devel/kodi-rockchip_18.2rc1-Leia/.armv8a-libreelec-linux-gnueabi && /var/lib/jenkins/LE/build2/workspace/box-trn9/build.LibreELEC-RK3328.arm-9.1-devel/toolchain/bin/cmake -DCORE_SOURCE_DIR=/var/lib/jenkins/LE/build2/workspace/box-trn9/build.LibreELEC-RK3328.arm-9.1-devel/kodi-rockchip_18.2rc1-Leia -DCORE_SYSTEM_NAME=linux -DCORE_BUILD_DIR=build -DCMAKE_BINARY_DIR=/var/lib/jenkins/LE/build2/workspace/box-trn9/build.LibreELEC-RK3328.arm-9.1-devel/kodi-rockchip_18.2rc1-Leia/.armv8a-libreelec-linux-gnueabi -DARCH_DEFINES="-D_LINUX -DTARGET_POSIX -DTARGET_LINUX -D_GNU_SOURCE -DHAVE_LINUX_MEMFD=1 -DHAVE_MKOSTEMP=1" -DAPP_SCMID=rockchip_18.2rc1-Leia -DAPP_COPYRIGHT_YEARS=2005-2018 -DAPP_BUILD_DATE=20190427 -Dprefix=/var/lib/jenkins/LE/build2/workspace/box-trn9/build.LibreELEC-RK3328.arm-9.1-devel/kodi-rockchip_18.2rc1-Leia/.armv8a-libreelec-linux-gnueabi/build -P /var/lib/jenkins/LE/build2/workspace/box-trn9/build.LibreELEC-RK3328.arm-9.1-devel/kodi-rockchip_18.2rc1-Leia/cmake/scripts/common/GenerateVersionedFiles.cmake
CMake Error at /var/lib/jenkins/LE/build2/workspace/box-trn9/build.LibreELEC-RK3328.arm-9.1-devel/kodi-rockchip_18.2rc1-Leia/cmake/scripts/common/GenerateVersionedFiles.cmake:28 (file):
  file COPY cannot copy file
  "/var/lib/jenkins/LE/build2/workspace/box-trn9/build.LibreELEC-RK3328.arm-9.1-devel/kodi-rockchip_18.2rc1-Leia/addons/xbmc.addon/repository.xsd"
  to
  "/var/lib/jenkins/LE/build2/workspace/box-trn9/build.LibreELEC-RK3328.arm-9.1-devel/kodi-rockchip_18.2rc1-Leia/.armv8a-libreelec-linux-gnueabi/addons/xbmc.addon/repository.xsd".


[9/1529] No patch step for 'dvdcss'
[10/1529] Copying files into build tree
```

LibreELEC is being clobbered by this repeatedly, which is truly annoying when Kodi is just a small part of the entire OS build and a random Kodi failure wastes hours of compute time.

I mentioned this on Team Kodi Slack, and was told the only solution is to repeat the Kodi build, but we don't special case build failures so that's not really an option - either the package (Kodi in this case) should build successfully, or it fails for a valid reason and not an annoying random bug.

I'm opening this issue in the hope that it might be fixed eventually without a hack workaround! Thanks.

**Comment by MilhouseVH** [comment]
I don't believe the template is relevant in the case of a build failure.

I should have posted the full Kodi build log though, so here it is: https://pastebin.com/raw/s2W4Nau8

**Comment by wsnipex** [comment]
please check if /var/lib/jenkins/LE/build2/workspace/box-trn9/build.LibreELEC-RK3328.arm-9.1-devel/kodi-rockchip_18.2rc1-Leia/.armv8a-libreelec-linux-gnueabi/addons/xbmc.addon exists when that happens.
Also, does it only happen with ninja?

**Comment by MilhouseVH** [comment]
Yes, the folder exists:

![s1](https://i.imgur.com/IA8dPCW.png)

> Also, does it only happen with ninja?

I don't believe this is ninja specific as it happens on Team Kodi Jenkins too:

![s2](https://i.imgur.com/2OikpJy.png)

I'll run some test builds using only cmake on Ubuntu 17.10 until I get a failure - unfortunately it's pretty random, so might take some time. We're seeing this failure quite frequently in our Jenkins but that's running a lot of builds (and is also more difficult to reconfigure for a test).

**Comment by kib** [comment]
I just noticed the picture is showing the 
workspace of box-trn9 on build5, but what was asked was 
workspace of box-trn9 on build**2** 

Is this the correct one ?

**Comment by MilhouseVH** [comment]
@kib I'm not sure what's going on with that Jenkins, to be honest - the log is from the same build5 workspace, but I don't know why the log refers to build2.

This is the image in Jenkins of the workspace where the log is located which clearly says build5:
![s1](https://i.imgur.com/2g3Xu3f.png)

and this is the contents of that stdout file, where it refers to build2:

http://ix.io/1Htc

Build #265 is the last build on this slave:
![s2](https://i.imgur.com/M0Y7jDH.png)

According to the environment variables all the build saves are configured to use build2 as the path... hmmm:
![s3](https://i.imgur.com/0LcEGI5.png)

I'll check with our Jenkins admin.

**Comment by kib** [comment]
Ok, I just happened to spot it and I wasn't sure whether it was relevant or just 'one of those things' :)

**Comment by islipfd19** [comment]
The build2 path is used commonly throughout each slaves configuration. Yes, it is confusing but is easiest to maintain for template and configuration file usage.

**Comment by MilhouseVH** [comment]
@kib - @islipfd19 is the LibreELEC Jenkins admin! :)

However we can take Jenkins out of the equation as I've been running repeated builds on a local server and have reproduced the failure wihtout Jenkins.

Log: http://ix.io/1Hx4

As you can see, it's a different file this time:
```
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.1-devel/kodi-18.2rc1-Leia/addons/screensaver.xbmc.builtin.dim/resources/language/resource.language.ko_kr/strings.po
```

and this file does exist:
```
ubuntu@buildbox:~/projects$ ls -la /home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.1-devel/kodi-18.2rc1-Leia/addons/screensaver.xbmc.builtin.dim/resources/language/resource.language.ko_kr/
total 12
drwxrwxr-x  2 ubuntu ubuntu 4096 Apr 29 05:04 .
drwxrwxr-x 62 ubuntu ubuntu 4096 Apr 29 05:04 ..
-rw-rw-r--  1 ubuntu ubuntu  698 Apr 29 05:04 strings.po
```

This is building LibreELEC which is using cmake+ninja, with `-j8`

I'm running a standard Kodi depends build on Ubuntu using just cmake (also `-j8`) and so far I've not had a failure, but if you see failures on Team Kodi Jenkins then presumably it should fail eventually and ninja isn't the trigger.

**Comment by MilhouseVH** [comment]
Amendment to previous comment:

The file exists in both the source and target:
```
CMake Error at /home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.1-devel/kodi-18.2rc1-Leia/cmake/scripts/common/GenerateVersionedFiles.cmake:28 (file):
  file COPY cannot copy file
  "/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.1-devel/kodi-18.2rc1-Leia/addons/screensaver.xbmc.builtin.dim/resources/language/resource.language.ko_kr/strings.po"
  to
  "/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.1-devel/kodi-18.2rc1-Leia/.x86_64-libreelec-linux-gnu/addons/screensaver.xbmc.builtin.dim/resources/language/resource.language.ko_kr/strings.po".
```

Source:
```
ubuntu@buildbox:~/projects$ ls -la /home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.1-devel/kodi-18.2rc1-Leia/addons/screensaver.xbmc.builtin.dim/resources/language/resource.language.ko_kr/strings.po
-rw-rw-r-- 1 ubuntu ubuntu 698 Apr 29 05:04 /home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.1-devel/kodi-18.2rc1-Leia/addons/screensaver.xbmc.builtin.dim/resources/language/resource.language.ko_kr/strings.po
```
Target:
```
ubuntu@buildbox:~/projects$ ls -la /home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.1-devel/kodi-18.2rc1-Leia/.x86_64-libreelec-linux-gnu/addons/screensaver.xbmc.builtin.dim/resources/language/resource.language.ko_kr/strings.po
-rw-rw-r-- 1 ubuntu ubuntu 698 Apr 29 05:04 /home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.1-devel/kodi-18.2rc1-Leia/.x86_64-libreelec-linux-gnu/addons/screensaver.xbmc.builtin.dim/resources/language/resource.language.ko_kr/strings.po
```

**Comment by MilhouseVH** [comment]
This is a `cmake` based failure: http://ix.io/1OQV
```
make[3]: Leaving directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu'
Scanning dependencies of target sse4
make[3]: Leaving directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu'
make[3]: Entering directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu'
[  0%] Generating build/xbmc/CompileInfo.cpp, addons/audioencoder.kodi.builtin.aac/addon.xml, addons/audioencoder.kodi.builtin.wma/addon.xml, addons/kodi.binary.global.audioengine/addon.xml, addons/kodi.binary.global.filesystem/addon.xml, addons/kodi.binary.global.general/addon.xml, addons/kodi.binary.global.gui/addon.xml, addons/kodi.binary.global.main/addon.xml, addons/kodi.binary.global.network/addon.xml, addons/kodi.binary.instance.audiodecoder/addon.xml, addons/kodi.binary.instance.audioencoder/addon.xml, addons/kodi.binary.instance.game/addon.xml, addons/kodi.binary.instance.imagedecoder/addon.xml, addons/kodi.binary.instance.inputstream/addon.xml, addons/kodi.binary.instance.peripheral/addon.xml, addons/kodi.binary.instance.pvr/addon.xml, addons/kodi.binary.instance.screensaver/addon.xml, addons/kodi.binary.instance.vfs/addon.xml, addons/kodi.binary.instance.videocodec/addon.xml, addons/kodi.binary.instance.visualization/addon.xml, addons/screensaver.xbmc.builtin.black/addon.xml, addons/screensaver.xbmc.builtin.dim/addon.xml, addons/xbmc.addon/addon.xml
Scanning dependencies of target export-files
make[3]: Entering directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu'
make[3]: Entering directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu'
make[3]: Leaving directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu'
[  0%] Creating directories for 'libcpluff'
make[3]: Entering directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu'
[  0%] Creating directories for 'dvdcss'
make[3]: Entering directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu'
make[3]: Entering directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu'
[  0%] Building CXX object build/platform/linux/sse4/CMakeFiles/sse4.dir/CopyFrame.cpp.o
[  0%] Copying files into build tree
[  0%] Built target crossguid
[  0%] Building C object build/cores/dll-loader/exports/CMakeFiles/wrapper.dir/wrapper.c.o
make[3]: Entering directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu'
CMake Error at /home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/cmake/scripts/common/GenerateVersionedFiles.cmake:28 (file):
  file COPY cannot set modification time on
  "/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu/addons/audioencoder.kodi.builtin.wma/icon.png"


make[3]: *** [CMakeFiles/compileinfo.dir/build.make:87: build/xbmc/CompileInfo.cpp] Error 1
make[3]: Leaving directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/kodi-18.3-Leia/.x86_64-libreelec-linux-gnu'
make[2]: *** [CMakeFiles/Makefile2:828: CMakeFiles/compileinfo.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
```

Using `ninja` makes this failure much more likely, but the fundamental issue remains even with `cmake` - presumably if this issue can be fixed, then `ninja` would also be fixed.

**Comment by MilhouseVH** [comment]
Another `cmake` fail: http://ix.io/1ORO

```
make[3]: Entering directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-RPi.arm-9.80-devel/kodi-18.3-Leia/.armv6zk-libreelec-linux-gnueabi'
[  1%] Generating ServiceDescription.h
make[3]: Entering directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-RPi.arm-9.80-devel/kodi-18.3-Leia/.armv6zk-libreelec-linux-gnueabi'
Scanning dependencies of target retroplayer_messages
make[3]: Leaving directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-RPi.arm-9.80-devel/kodi-18.3-Leia/.armv6zk-libreelec-linux-gnueabi'
make[3]: Entering directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-RPi.arm-9.80-devel/kodi-18.3-Leia/.armv6zk-libreelec-linux-gnueabi'
[  1%] Generating AddonModuleXbmcwsgi.i.cpp
CMake Error at /home/neil/projects/LibreELEC.tv/build.LibreELEC-RPi.arm-9.80-devel/kodi-18.3-Leia/cmake/scripts/common/GenerateVersionedFiles.cmake:28 (file):
  file COPY cannot copy file
  "/home/neil/projects/LibreELEC.tv/build.LibreELEC-RPi.arm-9.80-devel/kodi-18.3-Leia/addons/screensaver.xbmc.builtin.black/icon.png"
  to
  "/home/neil/projects/LibreELEC.tv/build.LibreELEC-RPi.arm-9.80-devel/kodi-18.3-Leia/.armv6zk-libreelec-linux-gnueabi/addons/screensaver.xbmc.builtin.black/icon.png".


make[3]: *** [CMakeFiles/compileinfo.dir/build.make:87: build/xbmc/CompileInfo.cpp] Error 1
make[3]: Leaving directory '/home/neil/projects/LibreELEC.tv/build.LibreELEC-RPi.arm-9.80-devel/kodi-18.3-Leia/.armv6zk-libreelec-linux-gnueabi'
```

That's two failures today. I previously tested the `cmake` build last week and had no failures after 1,000 builds. Maybe I'm having rotten luck today, or #16039 is somehow making this failure more likely (ignore the fact the log claims to be building 18.3-Leia as I'm building the latest Kodi 19 master and #16039 is the only change today).

**Comment by fuzzard** [comment]
The only thing that may be causing it could be would be the following in CMakeLists.txt

https://github.com/fuzzard/xbmc/blob/3afa61450971b3c2776dd0154de9372a4a167763/CMakeLists.txt#L303

```
${CMAKE_SOURCE_DIR}/cmake/treedata/${CORE_SYSTEM_NAME}/${CORE_PLATFORM_NAME_LC}/*.txt
```

Maybe try revert that and see if the issue persists.
I assume your building for rpi, so would be looking for cmake/treedata/linux/rpi/*.txt, but ive had plenty of jenkins kicks through the life of that PR without failing like your seeing.

As darwin_embedded is the only system with sub folders, i can look at what our other options are if its the cause.

**Comment by MilhouseVH** [comment]
I'm building for x86_64 and also RPi (see above two failures - first is x86_64, second is RPi). I'll see if I can recreate a test environment to compare with/without the Darwin change, but it takes days to acquire meaningful figures and despite testing 1,000 builds the other week it's not always possible to trigger the failure.

**Comment by fuzzard** [comment]
Fully understand. If removing that line changes anything, it may only be masking it, as you were obviously experiencing this well before that PR went in, but if its a way that can reproduce the error easier, then there is hope to replicate and find the root cause i guess.

**Comment by wsnipex** [comment]
please give https://github.com/wsnipex/xbmc/commit/2b3aaea95a5f761bcf8276aa9c413ff4225838ce a try. Not a proper fix, but might reveal if we run this task multiple times in parallel.

**Comment by MilhouseVH** [comment]
@wsnipex: This is a `cmake` fail with your branch: https://github.com/xbmc/xbmc/compare/master...wsnipex:heisenbug, building 1f51aecabf38f38ea69f07fd8cf013de4aca4f8c: http://ix.io/1OU5

**Comment by MilhouseVH** [comment]
@wsnipex another one: http://ix.io/1OUl (same conditions as previous comment).

**Comment by MilhouseVH** [comment]
Third one: http://ix.io/1OVh (same conditions as previous two comments).

**Comment by wsnipex** [comment]
so, this doesn't only happen in GenerateVersionedFiles, but also in other file(COPY..) paths. 
And my last assumption that this is a race because we somehow call the copy stuff multiple times seems false.
This surely looks like some arcane bug. Can you give some details about which file systems you use, are build and source dir on different disk/fs etc?

A rather desperate attempt is https://github.com/wsnipex/xbmc/commit/716d7d65f26ef76902a8794f080bdd5ed0e383e4

Doing the same for ExportFiles.cmake would be rather expensive, since it has >1300 copy operations and I'm not even sure using an external process helps.
Can you first try if you still get the error on GeneratedVersionedFiles with the above commit?

**Comment by MilhouseVH** [comment]
> This surely looks like some arcane bug.

Possibly, in `cmake` maybe? We use cmake-3.14.5 built from source.

> Can you give some details about which file systems you use, are build and source dir on different disk/fs etc?

Sure. Filesystem is ext4, and both build and source dir are on the same disk/fs. Build host is Ubuntu 17.04 (AMD FX-8350) although I've seen this on a completely different host (Ubuntu 16.04, Intel Haswell). Using `ninja` _seemed_ to make this problem more likely, but maybe it's just dumb luck.

We also use `ccache` during the build, not sure if that has any bearing on this issue.

> A rather desperate attempt is wsnipex@716d7d6

I'll give it a try and let you know, thanks!

**Comment by MilhouseVH** [comment]
250/250 passes with https://github.com/wsnipex/xbmc/commit/716d7d6, now trying another 250 (or maybe more if there are no failures). Unfortunately it's difficult to know if this has fixed it, or if I just haven't triggered it yet... :)

**Comment by wsnipex** [comment]
you can grep for "---------- GenerateVersionedFiles.cmake: Error" to check if a retry occurred.

**Comment by MilhouseVH** [comment]
I need to find a more reliable way to reproduce these failures.

I ran over 1,000 test builds with https://github.com/wsnipex/xbmc/commit/716d7d6 and didn't experience a single failure - not even a retry message.

I removed the commit, restarted the test and got to 600 builds without failure (`darwin`), and then started to realise this was getting silly... :)

So I started a separate concurrent build which is building into a completely different build directory (`Xarwin`, but within the same repository), and although it hadn't even got to building Kodi (it was still building the LibreELEC toolchain) the ongoing `darwin` test failed for the first time, after [621 builds](http://ix.io/1Pym).

So this makes me wonder if `cmake` is using temporary files (`/tmp`?) and somehow clashing/conflicting with another `cmake` package build? I'm pretty sure that the LibreELEC build system is not writing to `/tmp` in a way that is likely to conflict with `cmake` (in one place we use `mktemp`, which is safe), nor is it writing to any filesystem location outside of it's own build directory, so two (or more) concurrent LibreELEC builds should **NOT** be interfering with each other (in fact I regularly build multiple LibreELEC images which may explain why I'm hitting this Kodi issue more than others).

I guess there's a few avenues to investigate:

1) Leave the `darwin`/`Xarwin` test builds running and see if further Kodi failures occur
2) If more failures are seen with #1, repeat the `darwin`/`Xarwin` tests but with https://github.com/wsnipex/xbmc/commit/716d7d6
3) Run multiple "standard" Kodi Ubuntu builds (outside of LibreELEC, ie. "depends builds") and see if the `file COPY` failure can be triggered

**Comment by MilhouseVH** [comment]
Quick update:

> 1. Leave the darwin/Xarwin test builds running and see if further Kodi failures occur

I've now had 4 out of 250 failures of each build (8 failures/500 builds in total), this is a significantly higher failure rate than when building only a single build directory (0 failures/1000+ builds).

Moving on to item 2....

**Comment by wsnipex** [comment]
cmake does not use a common temp dir AFAIK, so I don't see how multiple parallel builds could clash. Maybe it's some internal timing issue, that's only triggered by heavy IO. Maybe going through an external process, like my patch does, works around it. So I'm curious on the results of step 2

**Comment by MilhouseVH** [comment]
Quick update - using https://github.com/wsnipex/xbmc/commit/716d7d6 I've had 336 builds of "Xarwin" and 362 builds of "Darwin" with 0 (zero) failures. My test is scheduled to run to 1,000 builds each which will take another 3 days or so (it's running at about 7 minutes per Kodi build). I've scheduled heavy IO loads during these builds and it hasn't been an issue (so far - unfortunately having zero failures doesn't guarantee the problem is fixed but fingers crossed - I hope this comment hasn't jinxed it!)

**Comment by wsnipex** [comment]
no retries either?
I wonder if opening a bug at cmake could track this down. But that would probably require a ton more compile work for you, so I'd absolutely understand if you don't want to.

**Comment by MilhouseVH** [comment]
No retries so far! This feels like a very obscure and hard to reproduce bug in code I'm not familiar with so trying to pin this down could be tricky/frustrating. I'd imagine the cmake developers would want a reproducible test case etc. which is entirely reasonable but unlikely from me especially given the nature of this issue. I'll do more testing to confirm if this problem exists with the standard Kodi build to try and eliminate LibreELEC as one more variable, but beyond that if your commit works then that might be the best fix for now.

**Comment by MilhouseVH** [comment]
2000 builds with wsnipex@716d7d6 (1000 each in two concurrent builds) and 0 (zero) failures. I did have failures in other repositories/build directories which didn't have this commit, which might be a good sign.

Tomorrow I'll move on to step 3, which is to try and reproduce a failure in a plain non-LibreELEC Kodi depends build (without wsnipex@716d7d6), most likely trying to build multiple Kodi instances at once.

**Comment by MilhouseVH** [comment]
I've been running 4 concurrent Kodi builds, 1000 times each, on the same Ubuntu 17.10/AMD FX-8350 (8-core) build host (without the "external" commit), and have had 10 `file COPY` build failures:

This is a summary of my build steps for "Kodi on Ubuntu":

1. Copy a pre-cloned kodi repository (kodi master from a few days ago) `/tmp/<build-name>` (where `<build-name>` is one of `darwin`, `Xarwin`, `Yarwin`, `Zarwin` - these are just random names to identify each build)
2. `export CCACHE_DIR=/home/neil/.ccache-<build-name> && ccache --clear` to ensure each build has it's own ccache (which is cleared before the first build)
3. `cd /tmp/<build-name>`
4. Perform the following steps once per loop to build Kodi:
```
rm -fr kodi-build
mkdir kodi-build
cd kodi-build

cmake -j8 .. -DENABLE_INTERNAL_FLATBUFFERS=1 -DCMAKE_INSTALL_PREFIX=/usr/local || exit
cmake --build . -- VERBOSE=1 -j8 || exit
```

These build steps were repeated 1000 times for each `<build-name>`.

The Kodi source and build directories are completely isolated from each other, and also isolated from all LibreELEC build activity, so any failures must be due to some sort of IO-related race condition. Or maybe `ccache` behaviour, even though each build has it's own 5GB CCACHE_DIR?

These were the 5 build failures (the digits represent the build sequence out of 1000):

darwin.969.log: http://ix.io/1R34 (`file COPY cannot copy file`)
Xarwin.2.log: http://ix.io/1R35 (`file COPY cannot copy file`)
Xarwin.696.log: http://ix.io/1R38 (`file COPY cannot set modification time on`)
Yarwin.888.log: http://ix.io/1R3d (`file COPY cannot copy file`)
Yarwin.973.log: http://ix.io/1R3h (`file COPY cannot copy file`)
Zarwin.245.log: http://ix.io/1R39 (`file COPY cannot copy file`)  
Zarwin.246.log: http://ix.io/1R3c (`file COPY cannot copy file`)
Zarwin.522.log: http://ix.io/1R3I (`file COPY cannot copy file`)
Zarwin.615.log: http://ix.io/1R5v (`file COPY cannot copy file`)
Zarwin.951.log: http://ix.io/1R8m (`file COPY cannot copy file`)


All failures occurred at different times (with one back-to-back failure Zarwin 245/246):
```
-rw-rw-r-- 1 neil neil  40283 Aug  6 01:43 Xarwin.2.log.gz
-rw-rw-r-- 1 neil neil  40737 Aug  7 21:31 Xarwin.696.log.gz
-rw-rw-r-- 1 neil neil  40873 Aug  7 22:05 Zarwin.245.log.gz
-rw-rw-r-- 1 neil neil  40884 Aug  7 22:08 Zarwin.246.log.gz
-rw-rw-r-- 1 neil neil  34717 Aug  8 10:26 Yarwin.888.log.gz
-rw-rw-r-- 1 neil neil  40586 Aug  8 17:49 darwin.969.log.gz
-rw-rw-r-- 1 neil neil  41184 Aug  8 18:00 Yarwin.973.log.gz
-rw-rw-r-- 1 neil neil  41022 Aug  8 19:30 Zarwin.522.log.gz
-rw-rw-r-- 1 neil neil  34573 Aug  9 00:52 Zarwin.615.log.gz
-rw-rw-r-- 1 neil neil  34596 Aug  9 17:01 Zarwin.951.log.gz
```

I'll now repeat the tests but this time *with* the "external" commit https://github.com/wsnipex/xbmc/commit/716d7d6

**Comment by wsnipex** [comment]
IMHO, you've already done more than enough testing. This must be a hard to trigger cmake bug in the file copy module and not something related to LE or kodi. 
I'm OK with using my external patch as workaround, but cannot prepare a PR atm(holiday), so feel free to do that whenever you want.

**Comment by MilhouseVH** [comment]
Thanks. I think I'll leave the final tests running with the "external" commit just to see what happens - it would be really nice if we can get to 4,000 builds without a single failure as that should give a higher level of confidence in the "external" commit, and assuming we can achieve that then I'll PR it (probably towards the end of next week). Obviously if we still get "COPY" failures then we may need to rethink this! :smile:

**Comment by MilhouseVH** [comment]
@wsnipex I've completed my testing, 1000 x 4 concurrent builds with zero failures (not even a retry).

I haven't ever seen a "retry" attempt in any of my testing when using the external copy method, so this would seem unnecessary and I'll shortly push a PR without the "retry" support. Many thanks for your assistance on this!
